### ISSUE TITLE
Select lambda for desired FDR level?

### ISSUE THREAD

**Issue by jarbet** [pbreheny/ncvreg#38]
**Is it possible to choose the value of`lambda` to get an expected FDR of level `alpha` for the selected features?**

Here is my R code, but I notice sometimes the estimated mFDR from `mfdr()` for a particular value of lambda can be very different than the "average mFDR" from `summary(fit, lambda)`, so I think I might be doing this wrong.  For example:

``` r
library(ncvreg);
library(ashr);
options(scipen=999);
alpha <- 0.20;

fit <- ncvreg(Heart$X, Heart$y, family="binomial");
fdr <- mfdr(fit);


fdr[fdr$mFDR <= alpha,];
#>                            EF S                  mFDR
#> 0.17746 0.0000000000000000000 0 0.0000000000000000000
#> 0.16550 0.0000000000006893422 1 0.0000000000006893422
#> 0.15435 0.0000000000268365145 1 0.0000000000268365145
#> 0.14394 0.0000000006329530280 1 0.0000000006329530280
#> 0.13424 0.0000000097763338989 1 0.0000000097763338989
#> 0.12519 0.0000001054851094933 1 0.0000001054851094933
#> 0.11676 0.0000008390208153253 1 0.0000008390208153253
#> 0.10889 0.0000051434436452641 1 0.0000051434436452641
#> 0.10155 0.0000253828646848793 2 0.0000126914323424397
#> 0.09470 0.0001034696509291972 2 0.0000517348254645986
#> 0.08832 0.0003551337306127198 2 0.0001775668653063599
#> 0.08237 0.0010615660523320986 4 0.0002653915130830247
#> 0.07682 0.0028082774755476881 4 0.0007020693688869220
#> 0.07164 0.0065715212877532712 4 0.0016428803219383178
#> 0.06681 0.0138633718317290964 4 0.0034658429579322741
#> 0.06231 0.0267721094431303992 4 0.0066930273607825998
#> 0.05811 0.0478096790345464359 5 0.0095619358069092865
#> 0.05419 0.0798471408546947287 5 0.0159694281709389443
#> 0.05054 0.1258959437651500091 5 0.0251791887530300011
#> 0.04713 0.1888348565257233214 5 0.0377669713051446629
#> 0.04396 0.2711688592920309948 5 0.0542337718584062004
#> 0.04100 0.3748380139080825790 5 0.0749676027816165158
#> 0.03823 0.5010544344203968770 5 0.1002108868840793782
#> 0.03566 0.6533316014300375452 5 0.1306663202860074979
#> 0.03325 0.8304030312135036729 5 0.1660806062427007401

lambda <- min(as.numeric(rownames(fdr[fdr$mFDR <= alpha,])));
lambda;
#> [1] 0.03325

summary(fit, lambda = lambda);
#> MCP-penalized logistic regression with n=462, p=9
#> At lambda=0.0333:
#> -------------------------------------------------
#>   Nonzero coefficients         :   5
#>   Expected nonzero coefficients:   0.21
#>   Average mfdr (5 features)    :   0.042
#> 
#>         Estimate     z      mfdr Selected
#> age      0.05244 6.188  < 0.0001        *
#> famhist  0.81913 4.162 0.0021476        *
#> tobacco  0.05757 3.289 0.0513593        *
#> typea    0.02513 3.186 0.0703252        *
#> ldl      0.11427 3.116 0.0861100        *
fdr[as.character(lambda),]
#>               EF S      mFDR
#> 0.03325 0.830403 5 0.1660806
```

Notice `mfdr()` says the estimated `mFDR = 0.166` for `lambda = 0.03325`.  However, the `summary` function says the average mFDR is 0.042.  So what is the expected FDR for this value of `lambda` ?

Is there a better way to select lambda for a desired level of FDR control?

**Comment by pbreheny** [comment]
Well, it can be shown that the expected value of the local marginal fdr (mfdr) is equal to the tail-area marginal fdr (mFDR). However,
* Sample averages are not equal to expected values, especially when taken over, say, only 5 features.
* The equality holds for the known values of mfdr and mFdr; in practice they must be estimated. Furthermore, mfdr is being estimated via the ashr package and mFdr is being estimated with a basic empirical CDF, introducing further disagreement between the two.

In other words, I don't find it terribly surprising that the average mfdr is 0.042 and the tail-area mFDR is 0.166. Both seem plausible given the uncertainty of estimates and small sample size (in terms of number of features) here.

**Comment by jarbet** [comment]
> Well, it can be shown that the expected value of the local marginal fdr (mfdr) is equal to the tail-area marginal fdr (mFDR). However,
> 
> * Sample averages are not equal to expected values, especially when taken over, say, only 5 features.
> * The equality holds for the known values of mfdr and mFdr; in practice they must be estimated. Furthermore, mfdr is being estimated via the ashr package and mFdr is being estimated with a basic empirical CDF, introducing further disagreement between the two.
> 
> In other words, I don't find it terribly surprising that the average mfdr is 0.042 and the tail-area mFDR is 0.166. Both seem plausible given the uncertainty of estimates and small sample size (in terms of number of features) here.

Gotcha.  So If I'm understanding, `mFDR` (from `mfdr()`) gives the expected FDR, but the "Average mfdr" from `summary`  gives a kind of sample version rather than the expected value.

So if I want to choose `lambda` to control the FDR at a desired level `alpha`, which of the following approaches would you recommend? Or would you suggest a different approach?

1. Use `mfdr()` and choose the minimum value of `lambda` such that `mFDR < alpha`
2. Use `summary` with a grid search over `lambda`.  Stop once you hit a value of `lambda` such that 1 or more of the features' `mfdr` > `alpha`, or stop once the "Average mfdr` > `alpha` (not sure which approach to use).

**Comment by pbreheny** [comment]
> mFDR (from mfdr()) gives the expected FDR, but the "Average mfdr" from summary gives a kind of sample version

Sort of, but not exactly. The true expected FDR would require knowing the true distribution of test statistics. Each approach has its own sources of divergence from this gold standard value:

1. mFDR/mfdr(): relies on Fhat to estimate F, the cdf
2. "Average mfdr" from summary: two sources (a) sample average (b) relies on fhat to estimate f, the density function

As far as what I would recommend -- choosing lambda via cross-validation, not mfdr. Personally, I don't find mfdr/mFDR a sensible way to choose a model, any more than I would recommend doing subset selection by p-values. The best models have some non-significant features in them; this is especially true for lasso models. This in no way makes them bad models.

**Comment by jarbet** [comment]
> > mFDR (from mfdr()) gives the expected FDR, but the "Average mfdr" from summary gives a kind of sample version
> 
> Sort of, but not exactly. The true expected FDR would require knowing the true distribution of test statistics. Each approach has its own sources of divergence from this gold standard value:
> 
> 1. mFDR/mfdr(): relies on Fhat to estimate F, the cdf
> 2. "Average mfdr" from summary: two sources (a) sample average (b) relies on fhat to estimate f, the density function
> 
> As far as what I would recommend -- choosing lambda via cross-validation, not mfdr. Personally, I don't find mfdr/mFDR a sensible way to choose a model, any more than I would recommend doing subset selection by p-values. The best models have some non-significant features in them; this is especially true for lasso models. This in no way makes them bad models.

Got it.  Thanks a lot for your replies!

**Comment by pbreheny** [comment]
Happy to help!
